# Questão 1

Calcule a fatoração LU, com pivoteamento parcial, da matriz

$
A =
\begin{bmatrix}
−0. & −10. & 13. & 2. \\
−16. & −2. & 5. & 5. \\
1. & −1. & 2. & −8. \\
−6. & −9. & −13. & −3.
\end{bmatrix}
$

In [1]:
import numpy as np

In [2]:
def FatoracaoLUPivot(A):
    U = np.copy(A)
    n = A.shape[0]
    L = np.zeros_like(A)
    P = np.eye(n)

    m = 0

    for j in range(n):

        #Pivoteamento Parcial

        k = np.argmax(np.abs(U[j:,j])) + j
        U[j], U[k] = np.copy(U[k]), np.copy(U[j])
        P[j], P[k] = np.copy(P[k]), np.copy(P[j])

        #print(L[2, 2:], L[3, 2:])
        L[j], L[k] = np.copy(L[k]), np.copy(L[j])
        L[j][j] = 1

        m += 1
        print(f"Linhas {j+1} e {k+1} trocadas")
        print(f"j = {j}: (linha {j+1})\nL = \n{L}\nU = \n{U}\nP = \n{P}\n\n")

        for i in range(j + 1, n):
            L[i][j] = U[i][j]/U[j][j]
            for k in range(j + 1, n):
                U[i][k] -= L[i][j] * U[j][k]
            U[i][j] = 0
            m += 1
            print(f"j = {j} & i = {i}: (termo ({i+1},{j+1}) zerado)\nL = \n{L}\nU = \n{U}\n\n")
    return L, U, P

In [3]:
A = np.array([[0., -10., 13., 2.],
[-16., -2., 5., 5.],
[1., -1., 2., -8.],
[-6., -9., -13., -3.]])

In [4]:
L, U, P = FatoracaoLUPivot(A)
L, U, P

Linhas 1 e 2 trocadas
j = 0: (linha 1)
L = 
[[1. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
U = 
[[-16.  -2.   5.   5.]
 [  0. -10.  13.   2.]
 [  1.  -1.   2.  -8.]
 [ -6.  -9. -13.  -3.]]
P = 
[[0. 1. 0. 0.]
 [1. 0. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


j = 0 & i = 1: (termo (2,1) zerado)
L = 
[[ 1.  0.  0.  0.]
 [-0.  0.  0.  0.]
 [ 0.  0.  0.  0.]
 [ 0.  0.  0.  0.]]
U = 
[[-16.  -2.   5.   5.]
 [  0. -10.  13.   2.]
 [  1.  -1.   2.  -8.]
 [ -6.  -9. -13.  -3.]]


j = 0 & i = 2: (termo (3,1) zerado)
L = 
[[ 1.      0.      0.      0.    ]
 [-0.      0.      0.      0.    ]
 [-0.0625  0.      0.      0.    ]
 [ 0.      0.      0.      0.    ]]
U = 
[[-16.      -2.       5.       5.    ]
 [  0.     -10.      13.       2.    ]
 [  0.      -1.125    2.3125  -7.6875]
 [ -6.      -9.     -13.      -3.    ]]


j = 0 & i = 3: (termo (4,1) zerado)
L = 
[[ 1.      0.      0.      0.    ]
 [-0.      0.      0.      0.    ]
 [-0.0625  0.      0.      0.    ]
 [ 0.375   0.      

(array([[ 1.        ,  0.        ,  0.        ,  0.        ],
        [-0.        ,  1.        ,  0.        ,  0.        ],
        [ 0.375     ,  0.825     ,  1.        ,  0.        ],
        [-0.0625    ,  0.1125    , -0.03320312,  1.        ]]),
 array([[-16.        ,  -2.        ,   5.        ,   5.        ],
        [  0.        , -10.        ,  13.        ,   2.        ],
        [  0.        ,   0.        , -25.6       ,  -6.525     ],
        [  0.        ,   0.        ,   0.        ,  -8.12915039]]),
 array([[0., 1., 0., 0.],
        [1., 0., 0., 0.],
        [0., 0., 0., 1.],
        [0., 0., 1., 0.]]))

In [5]:
print(P @ A, L @ U, sep="\n")

[[-16.  -2.   5.   5.]
 [  0. -10.  13.   2.]
 [ -6.  -9. -13.  -3.]
 [  1.  -1.   2.  -8.]]
[[-16.  -2.   5.   5.]
 [  0. -10.  13.   2.]
 [ -6.  -9. -13.  -3.]
 [  1.  -1.   2.  -8.]]


Aplicando os métodos estudados em aula, criei a rotina `FatoracaoLUPivot`. Ela recebe uma matriz $A$ e determina as matrizes $L$, $U$ e $P$ tal que $PA = LU$ em que $P$ é matriz de permutação, $L$ matriz triangular inferior com $1$s na diagonal principal e $U$ matriz diagonal superior. Ao longo do processo, a função esclarece quais operações foram realizadas e imprime as matrizes em estados intermediários.

Dessa forma, notei que foram trocadas as linhas 1 e 2, e 3 e 4 entre si, respectivamente. Calculando $PA$ e $LU$, comprovei que elas de fato são iguais e o algoritmo funcionou.

# Questão 2

Considere a matriz $M$ e o vetor $b$ definidos abaixo:

$
M =
\begin{bmatrix}
4. & −2. & 1. \\
1. & −1. & 1. \\
0. & 0. & 1. \\
1. & 1. & 1. \\
4. & 2. & 1. \\
\end{bmatrix}
$
e 
$
b =
\begin{bmatrix}
4. \\
10. \\
−5.
\end{bmatrix}
$
.

Determine a fatoração de Cholesky da matriz $A = M^TM$. Usando a fatoração de Cholesky, resolva o sistema linear $Ax = b$.

In [6]:
def SubstituicaoRegressiva(U, c): # U triangular superior
    x = np.copy(c)
    n = U.shape[0]

    for i in range(n-1, -1, -1):
        for j in range(i + 1, n):
            x[i] -= (U[i][j] * x[j])
        x[i] /= U[i][i]
    return x

In [7]:
def SubstituicaoDireta(U, c): #U triangular inferior
    x = np.copy(c)
    n = U.shape[0]

    for i in range(n):
        for j in range(i):
            x[i] -= (U[i][j] * x[j])
        x[i] /= U[i][i]
    return x

In [8]:
def FatoracaoCholesky(A):
    n = A.shape[0]
    G = np.zeros_like(A)

    for i in range(n):
        G[i,i] = np.sqrt(A[i,i] - np.sum(G[0:i,i] ** 2) )

        for j in range(i+1, n):
            G[i,j] = (A[i,j] - np.sum(G[0:i,j] * G[0:i, i]) ) / G[i,i]
    
    return G

In [9]:
def SolucaoCholesky(A, b):
    G = FatoracaoCholesky(A)

    # Resolver sistemas G^T y = b e G x = y

    y = SubstituicaoDireta(G.T, b)
    x = SubstituicaoRegressiva(G, y)
    return G, y, x

In [10]:
M = np.array([
[4., -2., 1.],
[1., -1., 1.],
[0., 0., 1.],
[1., 1., 1.],
[4., 2., 1.]])

In [11]:
b = np.array([
[4.],
[10.],
[-5.]])

In [12]:
print(M)
print(M.T)

[[ 4. -2.  1.]
 [ 1. -1.  1.]
 [ 0.  0.  1.]
 [ 1.  1.  1.]
 [ 4.  2.  1.]]
[[ 4.  1.  0.  1.  4.]
 [-2. -1.  0.  1.  2.]
 [ 1.  1.  1.  1.  1.]]


In [13]:
A = M.T @ M
A

array([[34.,  0., 10.],
       [ 0., 10.,  0.],
       [10.,  0.,  5.]])

In [14]:
G, y, x = SolucaoCholesky(A, b)

In [15]:
G

array([[5.83095189, 0.        , 1.71498585],
       [0.        , 3.16227766, 0.        ],
       [0.        , 0.        , 1.43486011]])

In [16]:
y

array([[ 0.68599434],
       [ 3.16227766],
       [-4.30458032]])

In [17]:
x

array([[ 1.],
       [ 1.],
       [-3.]])

In [18]:
print(M.T @ M @ x)
print(b)

[[ 4.]
 [10.]
 [-5.]]
[[ 4.]
 [10.]
 [-5.]]


Após implementar o algoritmo da Fatoração de Cholesky, e reutilizar códigos anteriores de substituições direta e regressiva, pude criar a função `SolucaoCholesky`. Ela, ao receber um sistema linear composto por uma matriz $A$ pressuposta simétrica e definida positiva e um vetor $b$, invoca a Fatoração de Cholesky em $A$ (obtendo $G$) e resolve o sistema ao decompô-lo

$
\begin{cases}
G^T y = b \\
G x = y
\end{cases}
$

e resolver as equações matriciais por substituições direta e regressiva, respectivamente.

A partir de $M$, calculei o vetor $A=M^TM$.

$
A =
\begin{bmatrix}
34 & 0 & 10 \\
0 & 10 & 0 \\
10 & 0 & 5
\end{bmatrix}
$

Então, utilizei `SolucaoCholesky` para obter o vetor $x$ do sistema $A x = b$.

O algoritmo encontrou $G$, $y$ e $x$.

$
G =
\begin{bmatrix}
5.83095189 & 0 & 1.71498585 \\
0 & 3.16227766 & 0 \\
0 & 0 & 1.43486011
\end{bmatrix}
$

$
y =
\begin{bmatrix}
0.68599434 \\
3.16227766 \\
-4.30458032
\end{bmatrix}
$

$
x =
\begin{bmatrix}
1 \\
1. \\
-3
\end{bmatrix}
$

Calculando $M^T M x = b$, pude observar que o programa resolveu o sistema linear corretamente.


